pkg/connector/account.go

```diff
@@ -139,6 +142,137 @@ func (a *accountBuilder) Grants(ctx context.Context, resource *v2.Resource, pTok
 	return rv, "", nil, nil
 }
 
+func (a *accountBuilder) Grant(ctx context.Context, principal *v2.Resource, entitlement *v2.Entitlement) (annotations.Annotations, error) {
+	l := ctxzap.Extract(ctx)
+
+	if !isValidPrincipal(principal.Id) {
+		l.Warn(
+			"databricks-connector: only users, groups and service principals can be granted account permissions",
+			zap.String("principal_id", principal.Id.String()),
+			zap.String("principal_type", principal.Id.ResourceType),
+		)
+
+		return nil, fmt.Errorf("databricks-connector: only users, groups and service principals can be granted account permissions")
+	}
+
+	accID := entitlement.Resource.Id.Resource
+	ruleSets, err := a.client.ListRuleSets(ctx, "", "")
+	if err != nil {
+		return nil, fmt.Errorf("databricks-connector: failed to list rule sets for account %s: %w", accID, err)
+	}
+
+	principalID, err := preparePrincipalID(ctx, a.client, principal.Id.ResourceType, principal.Id.Resource)
+	if err != nil {
+		return nil, fmt.Errorf("databricks-connector: failed to prepare principal id for principal %s: %w", principal.Id.Resource, err)
+	}
+
+	found := false
+
+	for i, ruleSet := range ruleSets {
+		if strings.Contains(ruleSet.Role, entitlement.Slug) {
+			found = true
+
+			// check if it contains the principals and add principal to the rule set
+			if slices.Contains(ruleSet.Principals, principalID) {
+				l.Info(
+					"databricks-connector: account already has the entitlement",
+					zap.String("principal_id", principalID),
+					zap.String("entitlement", entitlement.Slug),
+				)
+
+				return nil, nil
+			}
+
+			// add the principal to the rule set
+			ruleSets[i].Principals = append(ruleSets[i].Principals, principalID)
+		}
+	}
+
+	if len(ruleSets) == 0 || !found {
```

Comment:
If ruleSets is empty, found is always set to false.

```suggestion
	if !found {
```